### ISSUE TITLE
CP regression method: logistic & multinomial logistic regression

### ISSUE THREAD

**Issue by RichieHakim** [tensorly/tensorly#273]

**Comment by cohenjer** [comment]
@RichieHakim Thank you for sharing some ideas. With @caglayantuna  we are working on other losses at the moment, but only for the CP decomposition; maybe this will spark interest for others to implement CP regression with other losses?

**Comment by RichieHakim** [comment]
@cohenjer 
Thanks, I understand. The constrained CP that I see progress being made on looks awesome.
Do you have any tips on how to attempt trying it myself?
I'm not sure it would fit neatly into the LS method currently being used.

**Comment by cohenjer** [comment]
@RichieHakim So I would go about it as follows:
- Write your problem formally (in equations)
- Try to find a published research paper proposing an algorithm for what you are looking for. Maybe something along these lines (https://link.springer.com/chapter/10.1007/978-3-642-36669-7_70) ?
- If you find such a thing, then implement it in a new class, and even better you can use tensorly/lab to propose the feature
-  If you don't find existing works, and if you really need this feature, I can only suggest that you first try to publish a proposed method somewhere otherwise it will be difficult for other users to assess that your algorithm works.

**Comment by cohenjer** [comment]
The paper we are currently implementing is this one by the way: (https://arxiv.org/pdf/1808.07452.pdf) ; if it will satisfy your needs then you just need to wait a few months at worse :)

**Comment by RichieHakim** [comment]
I wrote one that works well. In case y'all are interested: https://github.com/RichieHakim/tensor_regression/blob/main/multinomial_tensor_regression.py

**Comment by JeanKossaifi** [comment]
@RichieHakim this is very interesting - it looks like it might be a good match for [TensorLy-Torch](https://github.com/tensorly/torch) as it's Pytorch only. Would you be interested in contributing it there?

**Comment by RichieHakim** [comment]
Sure, though I have never worked with Tensorly-Torch. The method is intended to be a multinomial logistic version of CPRegression in Tensorly. What would need to be done to contribute it to tl-torch? And what would need to be done to contribute it to tensorly proper?

**Comment by JeanKossaifi** [comment]
For adding to TensorLy-Torch it's mostly a matter of following the same conventions/format, trying to use TensorLy's existing functions where possible and having good unit-tests.

In TensorLy directly, all the functions/algorithms have to work transparently with any of the backends (including NumPy, PyTorch, TensorFlow, etc) - that means we can't use PyTorch only functions (in fact, that's why we created TensorLy-Torch). So concretely the way would be to write the gradients explicitly rather than rely on autograd, e.g. like Scikit-Learn does for instance.

**Comment by RichieHakim** [comment]
Ok, great. Is there somewhere you guys chat about details of development for this kind of stuff? I'm a bit green and have questions on how to integrate it properly and write unit-tests for it.
I don't know what the gradient for this problem is. If any of you do, I think it I would be happy to attempt using the unified `T` backend to try to get it to work in Tensorly. If not, I guess we'd have to make a new folder in tl-torch for full regression modules?

**Comment by JeanKossaifi** [comment]
@RichieHakim we have a Slack, you can join [here](https://join.slack.com/t/tensorly/shared_invite/zt-wqnts2sk-wbiRX6ml~Xt6~GDYWRPFfg )
